connectors/sources/servicenow.py

```diff
@@ -0,0 +1,346 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""ServiceNow source module responsible to fetch documents from ServiceNow."""
+import json
+import os
+
+import aiohttp
+from aiohttp.client_exceptions import ServerDisconnectedError
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource, ConfigurableFieldValueError
+from connectors.utils import CancellableSleeps, iso_utc
+
+RETRIES = 3
+RETRY_INTERVAL = 2
+FETCH_SIZE = 1000
+MAX_CONCURRENCY_SUPPORT = 10
+
+RUNNING_FTEST = (
+    "RUNNING_FTEST" in os.environ
+)  # Flag to check if a connector is run for ftest or not.
+
+ENDPOINTS = {
+    "TABLE": "/api/now/table/{table}",
+    "ATTACHMENT": "/api/now/attachment",
+    "DOWNLOAD": "/api/now/attachment/{sys_id}/file",
+}
+
+
+class InvalidResponse(Exception):
+    """Raise when retrieving invalid response..
+
+    Args:
+        Exception (InvalidResponse): Invalid response exception.
+    """
+
+    pass
+
+
+class ServiceNowClient(BaseDataSource):
```

Comment:
Why does the `ServiceNowClient` extend the `BaseDataSource`?

Reply:
I think it's better to just yield the docs from the client "raw" and serialize them in the `ServiceNowDataSource`. When we think of inheritance as a "is a" relationship it's a bit confusing IMO, when the client is also a data source.

Reply:
Could you elaborate a bit on the circular inheritance/concurrency problem with an example? Maybe we need to simply redesign the class structure to remove those issues. IMO we definitely shouldn't inherit from `BaseDataSource` in a client class.

Reply:
Faced this issue while implementing `get_docs` but for now we can remove this inheritance and will redesign class structure according to your suggestion

Reply:
Great, thanks!